src/system_database.ts

```diff
@@ -594,4 +607,55 @@ export class PostgresSystemDatabase implements SystemDatabase {
     };
     this.notificationsClient.on("notification", handler);
   }
+
+  /* SCHEDULER */
+  async getLastScheduledTime(wfn: string): Promise<number | null> {
+    const res = await this.pool.query<scheduler_state>(`
+      SELECT last_wf_sched_time
+      FROM ${DBOSExecutor.systemDBSchemaName}.scheduler_state
+      WHERE wf_function = $1;
+    `, [wfn]);
+
+    let v = res.rows[0]?.last_wf_sched_time ?? null;
+    if (v!== null) v = parseInt(`${v}`);
+    return v;
+  }
+
+  async setLastScheduledTime(wfn: string, invtime: number): Promise<number | null> {
+    const res = await this.pool.query<scheduler_state>(`
+      INSERT INTO ${DBOSExecutor.systemDBSchemaName}.scheduler_state (wf_function, last_wf_sched_time)
+      VALUES ($1, $2)
+      ON CONFLICT (wf_function)
+      DO UPDATE SET last_wf_sched_time = GREATEST(EXCLUDED.last_wf_sched_time, scheduler_state.last_wf_sched_time)
+      RETURNING last_wf_sched_time;
+    `, [wfn, invtime]);
+    
+    return parseInt(`${res.rows[0].last_wf_sched_time}`);
+  }
+
+  async getOutstandingScheduledWorkflows(wfn: string): Promise<number> {
+    const res = await this.pool.query<CountResult>(`
+      SELECT COUNT(scheduled_time) AS count
+      FROM ${DBOSExecutor.systemDBSchemaName}.scheduled_wf_running
+      WHERE wf_function = $1 AND scheduled_time IS NOT NULL;
+    `, [wfn]);
+    return parseInt(`${res.rows[0].count}`); // Returns the count of scheduled times
+  }
+  async scheduledWorkflowStarted(wfn: string, invtime: number): Promise<void> {
+    await this.pool.query(`
+      INSERT INTO ${DBOSExecutor.systemDBSchemaName}.scheduled_wf_running (wf_function, scheduled_time, actual_time)
+      VALUES ($1, $2, $3)
+      ON CONFLICT (wf_function, scheduled_time) DO NOTHING;
+    `, [wfn, `${invtime}`, `${new Date().getTime()}`]);
+  }
```

Comment:
This could have a comment saying: "if another node scheduled execution for this time slot, do nothing"